joeynmt/transformer.py

```diff
@@ -0,0 +1,233 @@
+# -*- coding: utf-8 -*-
+
+import math
+import copy
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+# This contains helper classes for the Transformer.
+# Source: http://nlp.seas.harvard.edu/2018/04/03/attention.html
+
+
+def clones(module, N):
+    """Produce N identical layers."""
+    return nn.ModuleList([copy.deepcopy(module) for _ in range(N)])
+
+
+# pylint: disable=arguments-differ,too-many-arguments
+class SublayerConnection(nn.Module):
+    """
+    Transformer residual connection
+    Applies sublayer to x, layer norm, dropout, then adds x.
+    """
+    def __init__(self, size, dropout):
+        super(SublayerConnection, self).__init__()
+        self.norm = nn.LayerNorm(size)
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, x, sublayer):
+        return x + self.dropout(sublayer(self.norm(x)))
+
+
+def subsequent_mask(size):
+    """
+    Mask out subsequent positions.
+    """
+    attn_shape = (1, size, size)
+    subseq_mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
+    return torch.from_numpy(subseq_mask) == 0
+
+
+def attention(query, key, value, mask=None, dropout=None):
+    """
+    Computes scaled dot-product attention
+    :param query:
+    :param key:
+    :param value:
+    :param mask:
+    :param dropout:
+    :return:
+    """
+    d_k = query.size(-1)
+    scores = torch.matmul(query, key.transpose(-2, -1)) \
+             / math.sqrt(d_k)
+    if mask is not None:
+        scores = scores.masked_fill(mask == 0, -1e9)
+    p_attn = F.softmax(scores, dim=-1)
+    if dropout is not None:
+        p_attn = dropout(p_attn)
+    return torch.matmul(p_attn, value), p_attn
+
+
+# pylint: disable=arguments-differ,too-many-arguments
+class MultiHeadedAttention(nn.Module):
+    def __init__(self, h, d_model, dropout=0.1):
+        """
+        Multi-headed attention
+        :param h: number of attention heads
+        :param d_model: model size
+        :param dropout:
+        """
+        super(MultiHeadedAttention, self).__init__()
+        assert d_model % h == 0
+        # We assume d_v always equals d_k
+        self.d_k = d_model // h
+        self.h = h
+        self.linears = clones(nn.Linear(d_model, d_model), 4)
```

Comment:
hard-coded 4: make configurable or specify as default value

Reply:
I think this is fine. The hard-coding is just to have an easy way to specify 4 linear layers.

Reply:
Now it's a default setting for the class argument.